### PR TITLE
79931 Log events

### PR DESCRIPTION
**Pull request by bratli**
DevOps#79931

Add history events, the same as for Preservation. Added ObjectGuid to IPO

### CODE REVIEW

File: src/Equinor.ProCoSys.IPO.Command/InvitationCommands/AcceptPunchOut/AcceptPunchOutCommandHandler.cs [hunk 1]
```diff
@@ -57,7 +57,7 @@ public async Task<Result<string>> Handle(AcceptPunchOutCommand request, Cancella
             }
             else
             {
-                AcceptIpoAsPersonAsync(invitation, participant, request.ParticipantRowVersion);
+                invitation.AcceptIpo(participant, participant.UserName, request.ParticipantRowVersion);
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Command/InvitationCommands/AcceptPunchOut/AcceptPunchOutCommandHandler.cs:60]: remove unused var e on line 74

File: src/Equinor.ProCoSys.IPO.Command/InvitationCommands/CompletePunchOut/CompletePunchOutCommandHandler.cs [hunk 2]
```diff
@@ -57,7 +57,7 @@ public async Task<Result<string>> Handle(CompletePunchOutCommand request, Cancel
             }
             else
             {
-                CompleteIpoAsPersonAsync(invitation, participants.SingleOrDefault(), request.ParticipantRowVersion);
+                invitation.CompleteIpo(participant, participant.UserName, request.ParticipantRowVersion);
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Command/InvitationCommands/CompletePunchOut/CompletePunchOutCommandHandler.cs:60]: remove unused var e on line 74

File: src/Equinor.ProCoSys.IPO.Infrastructure/EntityConfigurations/HistoryConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                .HasMaxLength(History.DescriptionLengthMax)
+                .IsRequired();
+
+            builder.Property(f => f.EventType)
+                .HasConversion<string>()
+                .IsRequired();
+
+            builder
+                .HasIndex(p => p.ObjectGuid)
+                .HasDatabaseName("IX_History_ObjectGuid_ASC");
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/EntityConfigurations/HistoryConfiguration.cs:28]: Any particular reason for custom name?

File: src/tests/Equinor.ProCoSys.IPO.Command.Tests/EventHandlers/EventDispatcherTests.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            var mediator = new Mock<IMediator>();
+            var dut = new EventDispatcher(mediator.Object);
+            var entities = new List<TestableEntityBase>();
+            for (var i = 0; i < 3; i++)
+            {
+                var entity = new Mock<TestableEntityBase>();
+                entity.Object.AddDomainEvent(new Mock<INotification>().Object);
+                entities.Add(entity.Object);
+            }
+            await dut.DispatchAsync(entities, default);
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Command.Tests/EventHandlers/EventDispatcherTests.cs:31]: remove default

File: src/tests/Equinor.ProCoSys.IPO.Command.Tests/EventHandlers/EventDispatcherTests.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Moq;
+
+namespace Equinor.ProCoSys.IPO.Command.Tests.EventHandlers
+{
+    [TestClass]
+    public class EventDispatcherTests
+    {
+        [TestMethod]
+        public void Constructor_ThrowsException_IFMediatorIsNull_Test() => Assert.ThrowsException<ArgumentNullException>(() => new EventDispatcher(null));
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Command.Tests/EventHandlers/EventDispatcherTests.cs:17]: long line

File: src/Equinor.ProCoSys.IPO.Infrastructure/EntityConfigurations/HistoryConfiguration.cs (hunk 3 above)
Comment (bratli) [reply src/Equinor.ProCoSys.IPO.Infrastructure/EntityConfigurations/HistoryConfiguration.cs:28]: No, just taken from preservation. Should I remove the "HasDatabaseName" ? (or all three lines?)

File: src/tests/Equinor.ProCoSys.IPO.Domain.Tests/AggregateModels/InvitationAggregate/InvitationTests.cs [hunk 6]
```diff
@@ -279,13 +318,136 @@ public void RemoveParticipant_ShouldThrowException_WhenParticipantNotGiven()
...
+        [TestMethod]
+        public void RemoveAttachment_ShouldRemoveAttachment()
+        {
+            Assert.AreEqual(1, _dutWithMcPkgScope.Attachments.Count);
+            Assert.AreEqual(_attachment, _dutWithMcPkgScope.Attachments.First());
+
+            _dutWithMcPkgScope.RemoveAttachment(_attachment);
+
+            Assert.AreEqual(0, _dutWithMcPkgScope.Attachments.Count);
+            Assert.IsInstanceOfType(_dutWithMcPkgScope.DomainEvents.Last(), typeof(AttachmentRemovedEvent));
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Domain.Tests/AggregateModels/InvitationAggregate/InvitationTests.cs:330]: I think the assert of correct type of event deserve its own test: RemoveAttachment_ShouldAddRemoveAttachmentEvent()
This goes for all asserts of same type in this testclass

**Review by janid1967 — APPROVED** [review]
Approved with suggestions

**Comment by bratli** [comment]
> Is it always true that Constructor should add the AttachmentUploadedEvent?
@janid1967 
I think so. Did you see any problems with this, or other times it should be added?

**Pull request merged**